File: packages/core/src/__tests__/MessageSerializer.test.ts
import { describe, it, expect } from 'vitest';
import { MessageSerializer } from '../message/MessageSerializer.js';

describe('MessageSerializer', () => {
  const serializer = new MessageSerializer('0.1.0');

  it('creates a request with correct fields', () => {
    const request = serializer.createRequest('device', 'getInfo', { key: 'value' });

    expect(request.id).toMatch(
      /^[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i,
    );
    expect(request.module).toBe('device');
    expect(request.action).toBe('getInfo');
    expect(request.payload).toEqual({ key: 'value' });
    expect(request.version).toBe('0.1.0');
  });

  it('defaults payload to empty object', () => {
    const request = serializer.createRequest('device', 'ping');
    expect(request.payload).toEqual({});
  });

  it('serializes a request to JSON', () => {
    const request = serializer.createRequest('storage', 'get', { key: 'foo' });
    const json = serializer.serialize(request);
    const parsed = JSON.parse(json);

    expect(parsed.id).toBe(request.id);
    expect(parsed.module).toBe('storage');
    expect(parsed.action).toBe('get');
    expect(parsed.payload).toEqual({ key: 'foo' });
    expect(parsed.version).toBe('0.1.0');
  });

  it('generates unique IDs', () => {
    const ids = new Set(
      Array.from({ length: 100 }, () => serializer.createRequest('m', 'a').id),
    );
    expect(ids.size).toBe(100);
  });
});

File: android/core/src/test/kotlin/com/devryner/rynbridge/core/MessageSerializerTest.kt
package com.devryner.rynbridge.core

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test

class MessageSerializerTest {

    @Test
    fun `createRequest generates UUID`() {
        val serializer = MessageSerializer(version = "0.1.0")
        val request = serializer.createRequest(module = "device", action = "getInfo")
        assertTrue(request.id.isNotEmpty())
        assertEquals("device", request.module)
        assertEquals("getInfo", request.action)
        assertEquals("0.1.0", request.version)
        assertTrue(request.payload.isEmpty())
    }

    @Test
    fun `createRequest with payload`() {
        val serializer = MessageSerializer()
        val request = serializer.createRequest(
            module = "storage",
            action = "set",
            payload = mapOf("key" to BridgeValue.string("test"), "value" to BridgeValue.string("hello"))
        )
        assertEquals("test", request.payload["key"]?.stringValue)
        assertEquals("hello", request.payload["value"]?.stringValue)
    }

    @Test
    fun `serialize request`() {
        val serializer = MessageSerializer(version = "0.1.0")
        val request = BridgeRequest(id = "test-id", module = "device", action = "getInfo", version = "0.1.0")
        val json = serializer.serialize(request)
        assertTrue(json.contains("\"id\":\"test-id\""))
        assertTrue(json.contains("\"module\":\"device\""))
        assertTrue(json.contains("\"action\":\"getInfo\""))
    }

    @Test
    fun `serialize response`() {
        val serializer = MessageSerializer()
        val response = BridgeResponse(
            id = "test-id",
            status = ResponseStatus.success,
            payload = mapOf("value" to BridgeValue.string("hello"))
        )
        val json = serializer.serialize(response)
        assertTrue(json.contains("\"id\":\"test-id\""))
        assertTrue(json.contains("\"status\":\"success\""))
    }

    @Test
    fun `each request gets unique ID`() {
        val serializer = MessageSerializer()
        val r1 = serializer.createRequest(module = "m", action = "a")
        val r2 = serializer.createRequest(module = "m", action = "a")
        assertNotEquals(r1.id, r2.id)
    }

    @Test
    fun `createResponse with error`() {
        val serializer = MessageSerializer()
        val errorData = BridgeErrorData(code = "TIMEOUT", message = "Request timed out")
        val response = serializer.createResponse(id = "r1", status = ResponseStatus.error, error = errorData)
        assertEquals(ResponseStatus.error, response.status)
        assertEquals("TIMEOUT", response.error?.code)
        val json = serializer.serialize(response)
        assertTrue(json.contains("TIMEOUT"))
    }
}
